### PR TITLE
adds error to redirect URL

### PR DESCRIPTION
**Pull request by ernie58**
In case of failure, adds an error parameter to the query string of the
failure redirect URL

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

File: lib/passport-configurator.js [hunk 1]
```diff
@@ -552,11 +555,20 @@ PassportConfigurator.prototype.configureProvider = function(name, options) {
...
         res.redirect(failureRedirect);
       });
   } else {
     var customCallback = options.customCallback || defaultCallback;
     // Register the path and the callback.
     self.app[callbackHTTPMethod](callbackPath, customCallback);
   }
+
+  function appendErrorToQueryString(url, err) {
+    return url + ((url.indexOf('?') === -1) ? '?' : '&') + 'error=' + encodeURIComponent(err);
```
Comment (superkhau) [inline lib/passport-configurator.js]: Please update this to something easier to read.

```
var hasExistingQueryString = url.indexOf('?') === -1;
var separator = hasExistingQueryString ? '?' : '&';
var fieldValuePair = 'error=' + encodeURIComponent(err);
var queryString = url + separator + fieldValuePair;
return queryString;
```

File: test/configure-errors.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
+(function() {
```
Comment (superkhau) [inline test/configure-errors.js]: Why do we need to function scope the whole file?

File: test/configure-errors.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  var loopback = require('loopback');
+  var app = loopback();
+  var m = require('./init');
+  var mock = require('mock-require');
+  var PassportConfigurator = require('../lib/passport-configurator.js');
+  var passportConfigurator = new PassportConfigurator(app);
```
Comment (superkhau) [inline test/configure-errors.js]: Do this inside the describe block at the beginning of the test during setup.

```
describe('...', function() {
   var passportConfigurator;
   
   before('set up passport configurator', function() {
     passportConfigurator = new PassportConfigurator(app);
   });
});
```

File: test/configure-errors.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  var passport = require('passport');
+  var assert = require('assert');
+  var request = require('supertest');
+  var sinon = require('sinon');
+  var passportAuthenticate;
```
Comment (superkhau) [inline test/configure-errors.js]: Put this var up with the ones inside the describe like above.

File: test/configure-errors.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+  //mock a strategy to work with
+  function PassportFacebookStrategy(options, verify) {};
+  PassportFacebookStrategy.prototype.authenticate = function(req, options) {};
+  mock('passport-facebook', PassportFacebookStrategy);
```
Comment (superkhau) [inline test/configure-errors.js]: Do all this inside a before block as described above. Add more as you need to describe what you're doing.

File: test/configure-errors.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+  describe('Login failure', function() {
+    before(function(done) {
```
Comment (superkhau) [inline test/configure-errors.js]: add a description to your before:

```
before('configure facebook provider', function() {
  ...
});
```

File: test/configure-errors.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        'link': false,
+      };
+      passportConfigurator.configureProvider('facebook-login', options);
+      done();
+    });
+
+    afterEach(function(done) {
+      if (passportAuthenticate) passportAuthenticate.restore();
+      done();
+    });
```
Comment (superkhau) [inline test/configure-errors.js]: move the aftereach above the actual tests

before() {}
afterEach() {}

it() {}

it() {}
```

Comment (superkhau) [inline test/configure-errors.js]: nevermind it already is, ignore this comment.

File: test/configure-errors.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        'failureQueryString': true,
+        'link': false,
+      };
+      passportConfigurator.configureProvider('facebook-login', options);
+      done();
+    });
+
+    afterEach(function(done) {
+      if (passportAuthenticate) passportAuthenticate.restore();
+      done();
```
Comment (superkhau) [inline test/configure-errors.js]: just make it sync and you don't need to call done

```
afterEach('reset stubbed functions', function() {
  if (passportAuthenticate)
    passportAuthenticate.restore();
});
```

File: test/configure-errors.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    });
+
+    it('redirects to failure with error', function(done) {
+      passportAuthenticate = sinon.stub(passport, 'authenticate',
+          getAuthenticateStubFunction('horrible mistake'));
```
Comment (superkhau) [inline test/configure-errors.js]: remove one level of indent:

```
passport...
  getAuthenticate...
```

File: test/configure-errors.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+      request(app)
+          .get('/auth/facebook/callback')
+          .expect(302)
+          .end(function(err, response) {
```
Comment (superkhau) [inline test/configure-errors.js]: nitpick: response can be shorted to res

File: test/configure-errors.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    it('redirects to failure without error if empty string', function(done) {
+      passportAuthenticate = sinon.stub(passport, 'authenticate', getAuthenticateStubFunction(''));
+      request(app)
+          .get('/auth/facebook/callback')
+          .expect(302)
+          .end(function(err, response) {
+            assert.equal(response.headers.location, '/auth/facebook-failure');
+            done();
+          });
+    });
```
Comment (superkhau) [inline test/configure-errors.js]: same things to fix from the previous comments

File: test/configure-errors.js [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+  });
+
+  describe('Link failure', function() {
+    before(function(done) {
```
Comment (superkhau) [inline test/configure-errors.js]: add description to before

File: test/configure-errors.js [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        'authPath': '/auth/facebook-link',
+        'callbackPath': '/auth/facebook-link/callback',
+        'successRedirect': '/auth/facebook-link/redirect',
+        'failureRedirect': '/auth/facebook-failure',
+        'failureQueryString': true,
+        'link': true,
+      };
+      var passportAuthorize = sinon.stub(passport, 'authorize', getAuthorizeStubFunction('omg'));
+      passportConfigurator.configureProvider('facebook-link', linkOptions);
+      passportAuthorize.restore();
```
Comment (superkhau) [inline test/configure-errors.js]: why do we stub passport authorize then restore it right after?

**Comment by superkhau** [comment]
@ernie58 Code reviewed, please update with my feedback.

File: test/configure-errors.js (hunk 13 above)
Comment (ernie58) [inline test/configure-errors.js]: passportConfigurator.configureProvider internally uses passport.authorize

**Comment by ernie58** [comment]
@superkhau I've updated the code

File: lib/passport-configurator.js [hunk 14]
```diff
@@ -552,11 +555,24 @@ PassportConfigurator.prototype.configureProvider = function(name, options) {
...
         res.redirect(failureRedirect);
       });
   } else {
     var customCallback = options.customCallback || defaultCallback;
     // Register the path and the callback.
     self.app[callbackHTTPMethod](callbackPath, customCallback);
   }
+
+  function appendErrorToQueryString(url, err) {
+    var noQueryStringYet = (url.indexOf('?') === -1);
```
Comment (superkhau) [inline lib/passport-configurator.js]: `var hasQueryString = (url.indexOf('?') === -1)`

File: lib/passport-configurator.js [hunk 15]
```diff
@@ -552,11 +555,24 @@ PassportConfigurator.prototype.configureProvider = function(name, options) {
...
+    var separator = (noQueryStringYet) ? '?' : '&';
```
Comment (superkhau) [inline lib/passport-configurator.js]: `var separator = hasQueryString ? '&' : '?';`

File: test/configure-errors.js [hunk 16]
```diff
@@ -0,0 +1,108 @@
+
```
Comment (superkhau) [inline test/configure-errors.js]: remove empty line

File: test/configure-errors.js [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+var loopback = require('loopback');
+var app = loopback();
```
Comment (superkhau) [inline test/configure-errors.js]: move this down into before block

File: test/configure-errors.js [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+var m = require('./init');
+var mock = require('mock-require');
+var PassportConfigurator = require('../lib/passport-configurator.js');
+var passport = require('passport');
+var assert = require('assert');
+var request = require('supertest');
+var sinon = require('sinon');
+
+describe('Login failure', function() {
```
Comment (superkhau) [inline test/configure-errors.js]: `var app;`

File: test/configure-errors.js [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+  var passportAuthenticate;
+
```
Comment (superkhau) [inline test/configure-errors.js]: add another before block:

```
before('set up loopback app', function() {
  app = loopback();
});
```
```

File: test/configure-errors.js [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+  before('Configure facebook strategy', function() {
```
Comment (superkhau) [inline test/configure-errors.js]: nitpick: lower case c in configure

File: test/configure-errors.js [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+      'callbackPath': '/auth/facebook/callback',
+      'successRedirect': '/auth/facebook/redirect',
+      'failureRedirect': '/auth/facebook-failure',
+      'failureQueryString': true,
+      'link': false,
+    };
+    passportConfigurator.configureProvider('facebook-login', options);
+  });
+
+  afterEach('Cleanup mocks', function() {
```
Comment (superkhau) [inline test/configure-errors.js]: same, here. we use all lowercase for descriptions

File: test/configure-errors.js [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+      .get('/auth/facebook/callback')
+      .expect(302)
+      .end(function(err, res) {
+        assert.equal(res.headers.location, '/auth/facebook-failure');
+        done();
+      });
+  });
+});
+
+describe('Link failure', function() {
```
Comment (superkhau) [inline test/configure-errors.js]: link

File: test/configure-errors.js [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+  before('Setup facebook strategy', function() {
```
Comment (superkhau) [inline test/configure-errors.js]: set up

**Comment by superkhau** [comment]
@ernie58 Reviewed, please update with the comments again. Almost there! ;)

File: lib/passport-configurator.js (hunk 14 above)
Comment (ernie58) [inline lib/passport-configurator.js]: If it has a queryString indexOf(?) should be larger than -1 ?

**Comment by ernie58** [comment]
Thx, I'll update tomorrow

File: lib/passport-configurator.js (hunk 14 above)
Comment (superkhau) [inline lib/passport-configurator.js]: Oops, my bad. `!== -1`

**Comment by ernie58** [comment]
@superkhau ping ^^^^^^

**Comment by ernie58** [comment]
@superkhau can you take a look at these as well? Thx

**Comment by superkhau** [comment]
LGTM. Thanks for the contribution! ;)

**Pull request merged**